### ISSUE TITLE
range.size can overflow

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#7530]
### Summary of Problem
The length of maximal uint ranges such as `0:uint(8)..255:uint(8)` overflows.
This is currently because the return type of length is the index type of the
range. You might be able to return `chpl__maxIntTypeSameSign(idxType)` for the
smaller uints, but for the `uint(64)` case, there's really nothing you can do
about it though.

This overflow is more annoying for int ranges where things like
`(0:int(8)..127..int(8)).length` result in overflow. Currently this case isn't
being tested in the .future because integer overflow is undefined.
I'm not sure if there's actually anything we can do about this issue over
having range.length returning uint(64)/int(64) instead of the index type but
that only masks the problem for some ranges.

I think I'd at least prefer some error checking that will report that overflow
is going to occur so the user isn't completely surprised by this.

**Associated Future Test(s):**
 [`test/types/range/elliot/rangeLength/rangeLengthOverflow.chpl`](https://github.com/chapel-lang/chapel/blob/master/test/types/range/elliot/rangeLength/rangeLengthOverflow.chpl)

**Comment by bradcray** [comment]
I think making `range.length` return `int` is a virtual no-brainer that isn't a perfect solution, but would avoid the most common cases.  Returning `uint` would be even more perfect, but is likely to run into standard int/uint mixing challenges.

At one point, we conceived of having ranges and domains be parameterized by a `capType` ("capacity type") which would be used as the return type for queries like `.length`, `.numIndices`, `.size` and the like, where this could default to `idxType` or `int`, and if a user wanted more capacity (a larger range/bigger domain), they'd have to dial it up to something bigger or `uint` explicitly and deal with the fallout (where a challenge would be making our module code robust to any capacity type).  I think for ranges we ended up thinking that this was too much complexity for a case that's unlikely to happen; for domains I think it continues to be an intriguing idea and think we just never got around to it (i.e., I don't recall failed attempts to implement it or anticipated challenges).

**Comment by vasslitvinov** [comment]
Here are our considerations on this topic from today's range module review. Spoiler: they echo [Brad's comment above](https://github.com/chapel-lang/chapel/issues/7530#issuecomment-333956276).

// Note: the `length` method is now called `size`.

It may look appealing to have `size` return the range's idxType because this way the returned value stays within the range algebra. However, we agreed that this valuable only in rare cases.

"There is no great return type for `size`." So, choose the type that works well in the majority of cases. Such type is `int`:
* It interoperates well with other code, avoiding int/uint mixing challenges.
* It is consistent with the return type of `size` on collections.
* It is sufficient in most cases.

To handle the cases where `int` is not suitable, we suggest adding another method, for example one where the user can request the desired return type.

// `int` is insufficient in more cases when invoking `size` on multi-dimensional domains.

Also: how should `size` and its variation behave when the return value overflows the return type? Ex. halt or throw an error; undefined behavior upon --fast?

**Comment by bradcray** [comment]
I was stewing on this some more last night, and like the approach:

```chapel
proc range.size: int { ... }
proc range.sizeAs(type t) { ... }
```

where we might either:

* have the `sizeAs()` routine default to `intIdxType`, `unsigned(intIdxType)` or `uint` 
* or not have it default to anything (which would make calls to it always read clearly, e.g., `myRange.sizeAs(uint)`

My current preference is for the second bullet, mostly because it supports readable code and avoids assumptions.  I'm also open to different names for this routine.

**Comment by bradcray** [comment]
In some spare moments this week, I got a draft of the previous comment's proposal mostly working with little effort.  As expected, it is a bit like pulling on a loose sweater thread, though, rippling pretty deep into the array code.

**Comment by bradcray** [comment]
I've quietly been running with this as a background task, so assigned myself.

**Comment by bradcray** [comment]
I've now got a working branch for this, though it needs cleanup and possibly extending before being ready to merge.

**Comment by bradcray** [comment]
I'm feeling like PR #17628 resolves this issue, but would want @ronawho to check that he agrees before I claim victory.

**Comment by ronawho** [comment]
I agree, this should be resolved by #17628
